### PR TITLE
NIFI-6782: Added repeat() String EL function

### PR DESCRIPTION
**Pull request by mattyb149**
Thank you for submitting a contribution to Apache NiFi.

Please provide a short description of the PR here:

#### Description of PR

_Adds the `repeat` function to NiFi Expression Language, for repeating a given string.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [x] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [x] Does your PR title start with **NIFI-XXXX** where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [x] Has your PR been rebased against the latest commit within the target branch (typically `master`)?

- [x] Is your initial contribution a single, squashed commit? _Additional commits in response to PR reviewer feedback should be made on this branch and pushed to allow change tracking. Do not `squash` or use `--force` when pushing to allow for clean monitoring of changes._

### For code changes:
- [x] Have you ensured that the full suite of tests is executed via `mvn -Pcontrib-check clean install` at the root `nifi` folder?
- [x] Have you written or updated unit tests to verify your changes?
- [x] Have you verified that the full build is successful on both JDK 8 and JDK 11?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the `LICENSE` file, including the main `LICENSE` file under `nifi-assembly`?
- [ ] If applicable, have you updated the `NOTICE` file, including the main `NOTICE` file found under `nifi-assembly`?
- [ ] If adding new Properties, have you added `.displayName` in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [x] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

**Comment by alopresto** [comment]
Matt, this looks awesome. Thanks for the quick turnaround. As soon as I finish PR 3821, I will review this.

**Comment by sburges** [comment]
I tested this change locally and it works!

The repeat function is really useful.

I'm not sure if this was discussed anywhere else or if I am missing something but I am not sure why the API has a second argument to repeat a random number of times. I was able to achieve the same result with an expression like this using just the single argument:

${"str":repeat( ${random():mod(10) } )}     //repeat a random number of times between 0 and 10
or
${"str":repeat( ${random():mod(10):plus(10)} )}   //repeat a random number of times between 10 adnd 20

This is also JMHO but the function "repeat(int)" makes a lot of sense. If a repeat random number of times, I would think something like repeatRandom(int) would be more clear

**Comment by mattyb149** [comment]
The 2-arg function is for convenience, the function chaining can be a bit cumbersome so I tried to combine them, although I might have fumbled over the documentation.

File: nifi-commons/nifi-expression-language/src/test/java/org/apache/nifi/attribute/expression/language/TestQuery.java [hunk 1]
```diff
@@ -2079,6 +2079,34 @@ public void testPadRight() {
         verifyEmpty("${nullString:padRight(10, \"@\")}", attributes);
     }
 
+    @Test
+    public void testRepeat() {
+        final Map<String, String> attributes = new HashMap<>();
+        attributes.put("str", "abc");
+
+        verifyEquals("${not_exist:repeat(1, 2)}", attributes, "");
+        verifyEquals("${str:repeat(1, 2)}", attributes, "abc");
```
Comment (alopresto) [inline nifi-commons/nifi-expression-language/src/test/java/org/apache/nifi/attribute/expression/language/TestQuery.java]: This line is not deterministic, as the potential results are `abc` and `abcabc`. I will fix when I merge.

Comment (mattyb149) [reply nifi-commons/nifi-expression-language/src/test/java/org/apache/nifi/attribute/expression/language/TestQuery.java]: Oops meant to change that when I made the `maxRepeats` inclusive. Good catch, and thanks for the fix (or I can certainly do it if you prefer)!

**Comment by alopresto** [comment]
I did a local build and tested this and I am experiencing odd behavior. While the code complete does suggest `repeat` and has the correct documentation, when I execute a `GenerateFlowFile` processor using it, I get `0 byte` content regardless of using the single or double parameter invocation. 

I've attached the sample flow I was using and a few screenshots. 
[flow.xml.gz](https://github.com/apache/nifi/files/3745465/flow.xml.gz)

generateflowfile configuration showing EL expression containing repeat

data provenance showing 0 byte filesize


log showing 0 byte filesize attribute and no flowfile content

**Comment by alopresto** [comment]
I also pushed the test fix to [4376d86](https://github.com/alopresto/nifi/commit/4376d86f6d6121976f918291a80d625f99aea2e7). Apparently my IDE did a lot of silent cleanup, but the only relevant change is line 2087+.

**Comment by alopresto** [comment]
I noticed that in the unit test, only attributes containing string values are tested. I am using a literal string in my smoke test. That may be the issue. I think all of the following should be supported. 

```
${"literal string":repeat(2)} -> "literal stringliteral string"
${literal('string_with_same_name_as_attr'):repeat(2)} -> "string_with_same_name_as_attrstring_with_same_name_as_attr"
${attr_name:repeat(2)} -> "attr_valueattr_value"
```

**Comment by ottobackwards** [comment]
Would anyone ever want spaces or separator between the repeats?  On the one hand, when would you be repeating a single value and care?  On the other hand:  string literalstring literal looks kind of strange.

Maybe a version with an optional separator parameter?

**Comment by alopresto** [comment]
I do think a custom delimiter string would be useful, but introducing one required and two (different) optional parameters gets tricky. I was assuming someone who cared could add a space intentionally or we might have to add a `join()`.

**Comment by ottobackwards** [comment]
It is tough to know without different people trying it whether a separator would be the 80% or the 20% case.

**Comment by spinachomes** [comment]
There is no repeat requirement so far

**Comment by mattyb149** [comment]
@alopresto Thanks for the test fix! I included it in my latest commit, hopefully GTG now

**Comment by MikeThomsen** [comment]
@mattyb149 you have a merge conflict now.

**Comment by mattyb149** [comment]
@alopresto @MikeThomsen fixed the merge conflict, hopefully good to go

**Pull request closed without merging**